### DISCUSSION TITLE
Proposal/Discussion: Syntactic sugar for constructor arguments and class level fields

### DISCUSSION THREAD

**Discussion by searus** [dotnet/csharplang#1087] [General]
It would be great if we could have some syntactic sugar similar to Typescript to specify class level fields for arguments passed into a constructor.  For example, this:
```
public class MyClass {

    private Dependency1 _dependency1;
    private Dependency2 _dependency2;

    public MyClass(
        Dependency1 dependency1, 
        Dependency2 dependency2
    ) {
        _dependency1 = dependency1;
        _dependency2 = dependency2;
    }

}
```
would become this:
```
public class MyClass {

    public MyClass(
        private Dependency1 _dependency1, 
        private Dependency2 _dependency2
    ) {
    }

}
```

Is this possible?

**Comment by Joe4evr** [comment]
Something similar was prototyped for C# 6 (Primary Constructor) before the team decided to delay that and rework it into Records (#39).

**Comment by Richiban** [comment]
@Joe4evr And it's still a massive disappointment to me that they did. After all, primary constructors have nothing to do with records (other than the fact that records automatically get a primary constructor). The only reason they did this as far as I can tell is because they wanted to steal the syntax `class Point(int x, int y)` for records.

I've copied a bit of code from a few classes within reach of me right now that show why I want such a feature so badly:

```csharp
    public sealed class CompleteCheckoutCommandHandler : ICompleteCheckoutCommandHandler
    {
        private readonly ICreateNewOrderForCheckoutCommandHandler _createNewOrderForCheckoutCommandHandler;
        private readonly ICommandSender _commandSender;
        private readonly ITakePaymentsCommandHandler _takePaymentsForOrderCommandHandler;
        private readonly IGetCustomerPaymentSourceQueryHandler _getCustomerPaymentSourceQueryHandler;

        public CompleteCheckoutCommandHandler(
            ICreateNewOrderForCheckoutCommandHandler createNewOrderForCheckoutCommandHandler,
            ICommandSender commandSender,
            ITakePaymentsCommandHandler takePaymentsForOrderCommandHandler,
            IGetCustomerPaymentSourceQueryHandler getCustomerPaymentSourceQueryHandler)
        {
            _createNewOrderForCheckoutCommandHandler = createNewOrderForCheckoutCommandHandler;
            _commandSender = commandSender;
            _takePaymentsForOrderCommandHandler = takePaymentsForOrderCommandHandler;
            _getCustomerPaymentSourceQueryHandler = getCustomerPaymentSourceQueryHandler;
        }

        // Methods removed
    }
```

```csharp

    internal sealed class DocumentDbBasketRepository : IBasketRepository
    {
        private readonly IDocumentDbBasketMapper _domainToDataDocumentDbBasketMapper;
        private readonly DocumentClient _basketClient;
        private readonly DocumentDBContextInfo _documentDbContextInfo;
        private readonly IBasketOwnerMapper _basketOwnerMapper;
        private readonly IBasketItemMapper _basketItemMapperToDomain;
        private readonly IGetProductImageUrlsQueryHandler _getProductImageUrlsQueryHandler;

        internal DocumentDbBasketRepository(
            DocumentClient basketClient,
            DocumentDBContextInfo documentDbContextInfo,
            IDocumentDbBasketMapper domainToDataDocumentDbBasketMapper,
            IBasketOwnerMapper basketOwnerMapper,
            IBasketItemMapper basketItemMapperToDomain,
            IGetProductImageUrlsQueryHandler getProductImageUrlsQueryHandler)
        {
            _basketClient = basketClient;
            _domainToDataDocumentDbBasketMapper = domainToDataDocumentDbBasketMapper;
            _basketOwnerMapper = basketOwnerMapper;
            _basketItemMapperToDomain = basketItemMapperToDomain;
            _getProductImageUrlsQueryHandler = getProductImageUrlsQueryHandler;
            _documentDbContextInfo = documentDbContextInfo;
        }

        // Methods removed
    }
```

```csharp
    public class PercentageValueDiscountService : IPercentageValueDiscountService
    {
        private readonly IDiscountRepository _discountRepository;

        public PercentageValueDiscountService(IDiscountRepository discountRepository)
        {
            _discountRepository = discountRepository;
        }

        // Methods removed
    }
```

Now, in our entire codebase _the vast majority of our classes_ look like this. It's incredibly frustrating to have to write these stupid constructors over and over again. 

Are we alone in this? Are we the only ones writing code where 90% of all constructors do nothing but set (private readonly) fields? To my eye they are overly verbose/repetitive, take up far too much space (obscuring the actual logic of your class which has only a single public method) and they're a hiding place for bugs.

I'd love to use an IL-weaving project such as Fody to just auto generate them all but I've tried it and it completely screws the design-time experience (ReSharper just doesn't understand what's going on at all and you get an awful lot of good-code-red).

Why did the team abandon the idea of primary constructors? Why can't I write (in the last example):

```csharp
    public class PercentageValueDiscountService(IDiscountRepository discountRepository) : IPercentageValueDiscountService
    {
        // Methods removed
    }
```

**NB** I know that records might enable a slightly shorter syntax for this, but these types **are not records**. They should have neither equality semantics nor public properties exposing these dependencies.

PS Sorry for the rant. I just _really_ want this.

**Comment by Richiban** [comment]
Copying from my comment on https://github.com/dotnet/csharplang/issues/39:

For reference, here are other languages which all support both primary constructors and records:

F#:
```fsharp
type Greeter(name : string) =
    member this.SayHello () = printfn "Hello, %s" name
```

Scala:

```scala
class Greeter(name: String) {
    def SayHi() = println("Hello, " + name)
}
```

Kotlin:

```kotlin
class Greeter(val name: String) {
    fun greet() {
        println("Hello, ${name}");
    }
}
```

**Comment by HaloFour** [comment]
@Richiban 

IIRC it was probably because the original conversation went down a path that was trying to turn primary constructors into an alternative syntax for full-fledged constructors complete with bizarre initialization and scoping syntax/rules.  By tying the feature to records it limits the use cases.

Why not use field injection?  It would eliminate the need for those constructors.  Or why not use quick-refactorings to write the constructors for you?  It's literally `Ctrl-.`, `<enter>`, `<enter>`.

That said I don't see a real reason for primary constructors to be necessarily limited to records, but I would hate to see it go down the same path it did before.  A succinct form of the syntax for limited but common scenarios would be good enough, and an additional `record` modifier could append the rest of the record behavior.

IIRC, all of your primary constructor examples in other languages result in publicly exposed fields, which is probably not what you want.

**Comment by Richiban** [comment]
@HaloFour 

> Why not use field injection?

Unfortunately that would significantly weaken the design; it would now be possible to construct these service types with null dependencies. As far as I'm aware no one recommends field injection any more and some IOC frameworks have even stopped supporting it.

> why not use quick-refactorings to write the constructors for you?

I am doing that currently. Writing the constructor isn't difficult, but reading and maintaining it are both an unreasonable burden.

> I would hate to see it go down the same path it did before. A succinct form of the syntax for limited but common scenarios would be good enough

I absolutely agree. I'm pretty sure that a syntax that simply declares and sets fields would already cover 95% of cases. It doesn't need to cover 100%: anyone with more custom needs is able to use the existing syntax.

> IIRC, all of your primary constructor examples in other languages result in publicly exposed fields

**Edit: verified all three examples**

No, they're private:
F#:
```fsharp
type Greeter(name : string) =
    member this.SayHello () = printfn "Hello, %s" name

let g = Greeter("Bob")

g.name // Error: The field, constructor or member 'name' is not defined
```

Scala:

```scala
class Greeter(name: String) {
    def SayHi() = println("Hello, " + name)
}

val g = new Greeter("Bob");

System.out.println(g.SayHi());
System.out.println(g.name); // Error: value name is not a member of Playground.this.Greeter
```

Kotlin:

```kotlin
class Greeter(val name: String) {
    fun greet() {
        println("Hello, ${name}");
    }
}

fun main(args: Array<String>) {
    val g = Greeter("Joe");

    System.out.println(g.greet())
    g.firstName             // Error: Unresolved reference: firstName
}
```

**Comment by jnm2** [comment]
Field injection? Yuck.

**Comment by Richiban** [comment]
Apologies, I made a mistake with the Kotlin example (I actually typed the wrong field name). @HaloFour was right and it does need a `private` annotation:

```kotlin
class Greeter(private val name: String) {
    fun greet() {
        println("Hello, ${name}");
    }
}

fun main(args: Array<String>) {
    val g = Greeter("Joe");

    System.out.println(g.greet())
    g.name             // Error: Cannot access 'name': it is private in 'Greeter'
}
```

**Comment by 0xF6** [comment]
Maybe:
```CSharp
public class Dog
{
    public [in] string Name;
}

...

var dog = new Dog("Hoffer");

dog.Name; // Hoffer
```
Although it is necessary to consider how to set the order...

```CSharp
public class Dog
{
    public [in] string Name;
    private [in] short Age;
    // or
    public [in] short Age { get; [private] or [protected] set; }
}

...

var dog = new Dog(Name: "Hoffer", Age: 4);

dog.Name; // Hoffer
dog.Age; // AccessException
```

What do you think?

**Comment by Pzixel** [comment]
@Richiban I don't like your example because you get inheritance list just being moved to another screen for large enough constructors. Consider your own example:
```csharp
    internal sealed class DocumentDbBasketRepository : IBasketRepository
    {
        private readonly IDocumentDbBasketMapper _domainToDataDocumentDbBasketMapper;
        private readonly DocumentClient _basketClient;
        private readonly DocumentDBContextInfo _documentDbContextInfo;
        private readonly IBasketOwnerMapper _basketOwnerMapper;
        private readonly IBasketItemMapper _basketItemMapperToDomain;
        private readonly IGetProductImageUrlsQueryHandler _getProductImageUrlsQueryHandler;

        internal DocumentDbBasketRepository(
            DocumentClient basketClient,
            DocumentDBContextInfo documentDbContextInfo,
            IDocumentDbBasketMapper domainToDataDocumentDbBasketMapper,
            IBasketOwnerMapper basketOwnerMapper,
            IBasketItemMapper basketItemMapperToDomain,
            IGetProductImageUrlsQueryHandler getProductImageUrlsQueryHandler)
        {
            _basketClient = basketClient;
            _domainToDataDocumentDbBasketMapper = domainToDataDocumentDbBasketMapper;
            _basketOwnerMapper = basketOwnerMapper;
            _basketItemMapperToDomain = basketItemMapperToDomain;
            _getProductImageUrlsQueryHandler = getProductImageUrlsQueryHandler;
            _documentDbContextInfo = documentDbContextInfo;
        }

        // Methods removed
    }
```
whould become
```csharp
    internal sealed class DocumentDbBasketRepository(
            DocumentClient basketClient,
            DocumentDBContextInfo documentDbContextInfo,
            IDocumentDbBasketMapper domainToDataDocumentDbBasketMapper,
            IBasketOwnerMapper basketOwnerMapper,
            IBasketItemMapper basketItemMapperToDomain,
            IGetProductImageUrlsQueryHandler getProductImageUrlsQueryHandler) : IBasketRepository
    {
        // Methods removed
    }
```
And when you have even more parameters (for complex enough type) it hides all base classes/interfaces.

I'd prefer something like:
```csharp
    internal sealed class DocumentDbBasketRepository : IBasketRepository
       with constructor(
            DocumentClient basketClient,
            DocumentDBContextInfo documentDbContextInfo,
            IDocumentDbBasketMapper domainToDataDocumentDbBasketMapper,
            IBasketOwnerMapper basketOwnerMapper,
            IBasketItemMapper basketItemMapperToDomain,
            IGetProductImageUrlsQueryHandler getProductImageUrlsQueryHandler)
    {        
    }
```

But I see here an issue with field naming. For example, should these fields have `_` prefix? Some teams use it, some don't, and we don't have an option to satisfy both.

**Comment by Richiban** [comment]
@Pzixel I don't think that the field naming convention is much of an issue; the underscore prefix is only a naming convention and it's far from universal. I foresee that most teams will switch naming convention if they have to for either the constructor parameters (little effect on code outside the class) or the fields (no effect outside the class).

As for subtypes being pushed down, again I'm fine with that. I don't think I've ever seen a class with a list of dependencies long enough to literally push the end of the parameters list off the screen (my standard 1080p screen can display about 45 lines). It might be reasonable to argue that with 40+ dependencies in a class you've got other problems :stuck_out_tongue_closed_eyes: I think it's obvious enough that the first line:

```csharp
internal sealed class DocumentDbBasketRepository(
``` 

is unfinished and reader's eye must continue before they should expect the list of subtypes.

**Comment by Pzixel** [comment]
@Richiban or [we merely could switch to F#](http://blog.ploeh.dk/2017/01/27/from-dependency-injection-to-dependency-rejection/) 😊

**Comment by cordasfilip** [comment]
@0xF6 I really like the syntax I think it's much more readable then putting anything in constructor definition. And it could simply translate from

```
public class Dog
{
    private in readonly string name;

       public Dog(string p1)
        {
              //Some constructor work
        }
}
```
to 
```
class Dog
    {
        private readonly string name;

        public Dog(string p1, string name)
        {
                this.name = name;
                 //Some constructor work
        }
    }
``` 
And inheritance could work with few caveats , if you had 
```
 class Dog: Animal { }

 class Animal
  {
      private in readonly string name;
  }
```
It would be the same as 
```
class Dog: Animal
    {
        public Dog(string name) : base(name)
        {
        }
    }

    class Animal
    {
        private readonly string name;

        public Animal(string name)
        {
                this.name = name;
        }
    }
```
Some naming rules should be defined like if there are two params with same name and optional params but this would make sense to me. You would get everything you get with constructor injection but with the nice syntax of property injection.

**Comment by jnm2** [comment]
It could be problematic because by adding or annotating a field you'd be creating breaking changes in the constructors without seeing it.

**Comment by HaloFour** [comment]
@0xF6 

I don't see how that would work with `partial` classes since each part of the class definition could define its own fields but they would have no specific order.

**Comment by cordasfilip** [comment]
@HaloFour Is this really such a big issue? I see this as a way to simplify a particular use case for DI. So disabling this for a partial class would not be that big of a deal. Or allowing only one part to use in fields in the class. It could also make two different constructors one for each part and let you deal with combining the two.

**Comment by MV10** [comment]
This is badly needed, along with addressing header-interfaces as in #1231...

At first glance I liked the `[in]` proposal, but I've changed my mind. I think it's important to retain the overt association with the constructor. It's beside the point, but I also don't think the word "in" clearly communicates what is happening. (All non-out parameters are in, right?)

I've noticed the C# designers prefer to avoid adding keywords. 

I propose reusing `var` with syntax similar to the C# 7.0 addition of inline declaration of `out` parameters.

```
public class Foo
{
    public Foo(var IBar bar);
}
```

The result is a `private readonly bar` since that is the most common pattern for DI. It's fast and easy to write, it's easy to understand at a glance, the developer still controls naming for those who prefer the Hungarian-like `_bar` convention, and it ought to be simple to implement.

It's probably obvious, but there's no reason to prevent mixing inline `var` with regular parameters, and I'm also proposing the constructor body as optional.

I also considered proposing access-visibility keywords, but I think that creates more problems than it solves (getters? setters? fields? readonly?). I think the private field approach is the most common pattern in use and creates the fewest opportunities for unexpected behaviors. In the rare case you really want to share the field, writing a public getter is easy enough.

**Comment by Richiban** [comment]
Here's another syntax, which I've half remembered from (I think) Dart? If you declare constructor parameters prefixed with a `this.` then it gets lifted into a field, e.g.

```csharp
public class Greeter
{
    public Greeter(string this.name) {}

    public string Greet()
        => $"Hello, {name}";
}
```

Having said that, there needs to be a bit more of a syntactic difference between this kind of constructor and normal constructors, because a few new rules come into force when you have a primary constructor. E.g. all other constructors must call this constructor.

What about this? It's always proved a bit of a syntactic papercut for me that you have to use the type name in a constructor:

```csharp
public class Greeter
{
    public new(string this.name) {}

    public string Greet()
        => $"Hello, {name}";
}
```

We're having to dance around a little bit since the best syntax: 

```csharp
public class Greeter(string name)
{
    public string Greet() => $"Hello, {name}";
}
```

has been stolen by records.

**Comment by MV10** [comment]
@Richiban 
Even better. That's very reminiscent of the way I write DI constructors today:

```
private readonly IFoo foo;

public MyClass(IFoo foo)
{
    this.foo = foo;
}
```

Funny you'd mention constructor names, just this morning I opened (and quickly closed) [this proposal](https://github.com/dotnet/csharplang/issues/1335). I like the `new` idea.

**Comment by Richiban** [comment]
@MV10 

Yeah, `new` for constructors is nice, I think. There's a potential source of confusion since `new` is valid as a keyword on methods, meaning that there are two very different member declarations that would have very similar syntaxes:

A new-style constructor:

```csharp
public class MyClass
{
    public new(string x, string y)
    {
        // ...
    }
}
```

A method called `M()` that returns a `(string x, string y)` tuple and shadows any base member with the same name.

```csharp
public class MyClass
{
    public new(string x, string y) M() 
    {
        // ...
    }
}
```

I'm not convinced this is going to be a big problem though; it's a bit of a contrived example.

**Comment by cordasfilip** [comment]
Here is something else that might help with this and probably would not be that hard to implement.
Something I always liked are _Object Initializers_ but they become unusable when working with immutable objects and there is no way to force the user of the class to set the property. If we use @0xF6 syntax for in property, we could indicate that the property must be used in Object Initializer.
So if we define a class like this. 
```
public class Test
{
    public in string Prop1 { get; set; }

    private in string Prop2 { get; set; }

    public in string Prop3 { get; private set; }

    public in string Prop4 { get; }
}

static void Main(string[] args)
{
    var t1 = new Test(); // Compile time error
    var t2 = new Test
    {
        Prop1 = "",
        Prop2 = "",
        Prop3 = "",
        Prop4 = ""
    };// No error even if we have private fields

}
```
To me this solves two problems with one solution. Makes object initializers work for immutable classes and DI injection could easily fill only the `in` properties. We would get compile time errors if no property is set and the build will fail if someone change something. This doesn't need to be run-time error really and if someone dose something he was not suppose to let him handle it on his own risk.  Since this are just properties everything will work as it dose now with inheritance and  partial...
And also some other way to let constructors set properties can still be implemented like new.

**Comment by Pzixel** [comment]
Allowing access private fields is the greatest evil. It's constructor job and this is how it should work.

I agree that we need something to deal with immutable types, but jeez, it's just like curing headash via guillotine.

**Comment by cordasfilip** [comment]
@Pzixel You are right about the constructor, but what I noticed is that I generally stop using immutable structures or not use object initializers. And I like both, since you can only do this during construction private could theoretically work, BUT I'm not saying this has to work with private I would be happy 😊 if it only worked with public properties. The private part can be dropped if people don't think it's useful i just think this makes more sense, but I don't think this would work without readonly properties (Prop4 in the example). This is just an idea to get people to think about it.

**Comment by Pzixel** [comment]
But if it's applicable to public properties only it becomes almost useless. You can see an example above, in most cases I ever seen you init private fields of class because otherwise you can get this value changed. For example if you get list of something as parameter you don't want expose it because someone may modify it. Make all properties completely immutable? It may be not applicable always and it also makes your codebase 2x larger because there is built in `ImmutableList`/`ImmutableArray`/..., but there is no `ImmutableMyBusinessLogicDto`

**Comment by jnm2** [comment]
> `ImmutableMyBusinessLogicDto`

When I write mine that way, the `Immutable` part is implied. 😁

**Comment by Pzixel** [comment]
@jnm2 hmm, why then my `DTO`s are always dumb objects with all properties being public with `DataMember` attribute? 😃

**Comment by jnm2** [comment]
It's not my fault they're dumb... 😉

**Comment by cordasfilip** [comment]
@Pzixel 

> But if it's applicable to public properties only it becomes almost useless.

 Not necessarily this will become useful with  [Non-Nullable](https://gist.github.com/olmobrutall/31d2abafe0b21b017d56) types. It will be simpler to force someone to put a value and not use default. I am still not sure how this will work when they implement non-nullable.

>  I ever seen you init private fields of class because otherwise you can get this value changed

Making the field readonly makes sure the value won't change, it can be public or private. The "immutable" property is the Prop4 in the example. But in regards to private properties, to me `private in` is different then `private` like` internal protected` is different from `protected ` and `internal`.
Also this will answer all the problems of [Property Injection
](https://stackoverflow.com/questions/29915192/property-injection-on-attributes) and constructor injection when doing inheritance.

**Comment by Pzixel** [comment]
@cordasfilip 

> Making the field readonly makes sure the value won't change

Making `List` readonly won't prohibit someone from modifying it.

**Comment by searus** [comment]
When I initially created this issue I actually thought it would turn out simpler than this 😊
As a daily user of TypeScript, I still think that the way they implement this would be the simplest.  Just prefix the constructor argument with "private" and have a readonly field created with the same name as the argument and the value set.

@ahejlsberg do you have any thoughts on this?

**Comment by MV10** [comment]
I still think Richiban had the best solution.

`public new(string this.name);`

Concise, obvious, no likely confusion with other syntax, the constructor is still a constructor doing constructor things, no new keywords required, minimal new syntax with no drastic re-interpretations around it, and the behind-the-scenes part is exactly like automatic backing variables for properties, which was acceptable to just about everyone. I see no downside.

**Comment by Richiban** [comment]
@searus 

I hadn't thought of Typescript! Good idea; here's an example:

```typescript
class Greeter
{
    constructor(private name: string) { }

    sayHello() { return `Hello, ${this.name}`; }
}
```

Looks pretty nice, IMO. It also solves the problem of interface implementation declarations being pushed down the page that @Pzixel mentioned (https://github.com/dotnet/csharplang/issues/1087#issuecomment-346857307).

**Comment by MattDillard** [comment]
I like the TypeScript approach. However, there are some problems with lifting that syntax directly into C#.

One major difference is that TypeScript permits only one constructor per class; not so with C#.

How would you propose a syntax like that working in the presence of multiple constructors, each of which may want to assign a value to that field? Things get even trickier if an "extension everything" proposal includes extension constructors. I don't believe anyone has an appetite for extension members adding state to a class.

**Comment by Richiban** [comment]
That's why it might be an idea to have these primary constructors look a bit different to ordinary ones.

If we take a leaf out of F#''s book, once fields are defined the constructor becomes a _primary constructor_ and all other constructors must (ultimately) call the primary one.

**Comment by liguori** [comment]
We need for the next release C# 8, syntactic sugar that will be a mix between:
-Primary Constructors  - Really disappointed that was removed from C# 6, no relation with tuple ;)
-Constructor Parameter Properties- It's working very good in the various TypeScript codebase

**Comment by MrBildo** [comment]
I know this issue is old, but it's still relevant. I'm just looking for an implied constructor (like how a default parameter-less constructor is implied) for derived classes.

```
    public abstract class BaseClass
    {
        public BaseClass(string parameter, int anotherParameter)
        {
            //...
        }
    }

    public class ChildClass : BaseClass
    {
        //no constructor defined
        
        //use implied constructor with the same signature as base

    }

    public class TestItOut
    {
        void SomeMethod()
        {
            var child = new ChildClass("parameter", 100); //this works!
        }
    }
```

**Reply by CyrusNajmabadi** [reply]
You could potentially do this with a source-generator.  Just write one that looks for derived types that are missing a constructor, and derive from a base without an empty constructor.  In such a case, it could generate this constructor for you.

**Comment by Pzixel** [comment]
Let's put it in different way: what you *can't* write with source generators? I'm pretty sure that with generators of sufficient complexity you can implement any feature 😃

**Reply by CyrusNajmabadi** [reply]
> Let's put it in different way: what you can't write with source generators? 

You can't change the bodies of methods (or any existing code for that matter).  You can only do additive things.  As this is a purely additive language request, it's very suited to just using a SourceGenerator here.

Put another way:  the feature to do this already exists and can be used today.  It feels unnecessary to add a language feature given that.
